xmtp_mls/README.md

```diff
@@ -0,0 +1,143 @@
+    created_at_ns BIGINT NOT NULL, -- Based on the timestamp of the welcome message
+    membership_state INT NOT NULL, -- Enum of the GROUP_MEMBERSHIP_STATE listed below
+}
+
+-- Successfully processed messages meant to be returned to the user
+CREATE TABLE group_messages {
+    id BLOB PRIMARY KEY NOT NULL -- Derived via SHA256(CONCAT(decrypted_message_bytes, conversation_id, timestamp))
+    group_id BLOB NOT NULL,
+    decrypted_message_bytes BLOB NOT NULL, -- Message contents after decryption
+    sent_at_ns BIGINT NOT NULL, -- Based on the timestamp of the message
+    sender_installation_id TEXT NOT NULL,
+    sender_wallet_address TEXT NOT NULL, -- Could remove this if we added a table mapping installation_ids to wallet addresses
+}
+
+-- Used to keep track of the last seen message timestamp in a topic
+CREATE TABLE topic_refresh_state {
+    topic TEXT PRIMARY KEY NOT NULL,
+    last_message_timestamp BIGINT NOT NULL
+}
+
+-- This table is required to retry messages that do not send successfully due to epoch conflicts
+CREATE TABLE group_intents {
+    id SERIAL PRIMARY KEY NOT NULL,
+    kind INT NOT NULL, -- INTENT_KIND
+    group_id BLOB NOT NULL,
+    "data" BLOB NOT NULL, -- Some sort of serializable blob that can be used to re-try the message if the first attempt failed due to conflict
+    "state" INT NOT NULL, -- INTENT_STATE,
+    message_hash BLOB -- The hash of the encrypted, concrete, form of the message if it was published.
+}
+
+CREATE TABLE outbound_welcome_messages {
+    id BLOB PRIMARY KEY NOT NULL, -- Derived via SHA256(CONCAT(group_id, welcome_message, ratchet_tree, installation_id))
+    state INT NOT NULL, -- OUTBOUND_WELCOME_STATE
+    installation_id BLOB PRIMARY KEY NOT NULL,
+    commit_hash BLOB NOT NULL, -- The hash of the commit which created this
+    group_id BLOB NOT NULL,
+    welcome_message BLOB NOT NULL,
+    ratchet_tree BLOB NOT NULL,
+}
+
+-- OpenMLS Keystore backing database to lookup objects. Interface and matching types defined here https://github.com/openmls/openmls/blob/main/traits/src/key_store.rs
+CREATE TABLE openmls_keystore {
+    "key" BLOB PRIMARY KEY NOT NULL,
+    "value" BLOB NOT NULL,
+}
+```
+
+## Enums
+
+### GROUP_MEMBERSHIP_STATE
+
+- ALLOWED // User has agreed to be a member of the group
+- REJECTED // User has rejected an invite to the group or left
+- PENDING // User has neither accepted or rejected whether they should join the group
+
+### INTENT_STATE
+
+- TO_SEND // Either has never been sent to the network or needs to be re-sent
+- PUBLISHED // Sent to the network but has not been read back or committed
+- COMMITTED // Committed messages could be deleted
+
+### INTENT_KIND
+
+- SEND_MESSAGE // An intent to send a message to the group
+- ADD_MEMBERS // An intent to add members to the group
+- REMOVE_MEMBERS // An intent to remove members from the group
+- KEY_UPDATE // An intent to update your own group key
```

Comment:
I would propose these intent kinds:
- SEND_MESSAGE: Send a message to members of the group
- COMMIT_NEW: Send a commit covering new actions (update / remove / add) to the group
- COMMIT_PENDING: Send a commit covering pending proposals

A "COMMIT_PENDING" intent is created when the application is unable to send an application message because there are pending proposals. Then you can specify that intents are always processed in this order:
- COMMIT_PENDING intents always get executed first
- Then SEND_MESSAGE and COMMIT_NEW intents are executed in the order they were inserted in the table (note: requires adding timestamp column to group_intents)

It is probably easiest for now to handle each intent as blocking; you can't handle the second one until the first one is complete and deleted.

Reply:
If we have a generic "commit" intent, my concern is how hard it would be to untangle the commit and reconstruct it in a later epoch. Do we still have access to the Key Packages of new members etc in a format that is usable for generating a new commit? I'd have to dig in a little deeper to OpenMLS to figure out just how accessible that information is.

Reply:
> Then SEND_MESSAGE and COMMIT_NEW intents are executed in the order they were inserted in the table (note: requires adding timestamp column to group_intents)

I was imagining using the SERIAL primary key to handle sequencing, but I should have called that out more clearly in the doc.

> It is probably easiest for now to handle each intent as blocking; you can't handle the second one until the first one is complete and deleted.

That feels like a reasonable place to start

Reply:
> COMMIT_PENDING: Send a commit covering pending proposals

@Bren2010 I didn't understand what this means. What happens when this intent is processed?

> A "COMMIT_PENDING" intent is created when the application is unable to send an application message because there are pending proposals

Can't we simply store an `intent` with the plaintext message, and then later on send the message after all previous intents have been executed?

Reply:
Oh I think I understood your intention here. The current idea is to not use proposals at all, only commits. Are there drawbacks to this?

Reply:
Summary of chat: We can choose to send proposals and commits in the same payload, but in order to be standards-compliant we can't assume this is the case when receiving. Therefore, we may have pending proposals at any time. This will in turn block the sending of application messages

Reply:
> A "COMMIT_PENDING" intent is created when the application is unable to send an application message because there are pending proposals

What problem arises if we simply commit pending proposals as part of the `send message` intent (and possibly also every other intent)?

Reply:
> If we have a generic "commit" intent, my concern is how hard it would be to untangle the commit and reconstruct it in a later epoch.

The reason I proposed a generic "commit" intent was to allow actions of several different types to be batched together. I see below that you're scoping that out for now, so this should be fine!